### PR TITLE
Packetfilter: add support for namedset

### PR DESCRIPTION
**Pull request by yboaron**
Add support for generic namedSet APIs under packetfilter.

### CODE REVIEW

File: pkg/packetfilter/packetfilter.go [hunk 1]
```diff
@@ -138,6 +136,37 @@ type ChainIPHook struct {
...
+	// SetType specifies the named type.
+	SetType string
+	// nftables named set attached to tables
+	Table TableType
+	// SetFamily specifies the protocol family of the IP addresses to be stored in the set.
+	// The default is IPv4.
+	Family SetFamily
+}
+
+type Named interface {
```
Comment (tpantelis) [inline pkg/packetfilter/packetfilter.go]: I think this should be called `NamedSet` as this is what consumers will store and use continuously.  The struct above is referenced once as input so I suggest `SetConfig` or `SetInfo`.

```suggestion
type NamedSet interface {
```

File: pkg/packetfilter/packetfilter.go [hunk 2]
```diff
@@ -153,6 +182,11 @@ type Driver interface {
 	List(table TableType, chain string) ([]*Rule, error)
 	Append(table TableType, chain string, rule *Rule) error
 	Insert(table TableType, chain string, pos int, rule *Rule) error
+
+	// named Sets
+	NewNamed(set *NamedSet) Named
```
Comment (tpantelis) [inline pkg/packetfilter/packetfilter.go]:
```suggestion
	NewNamedSet(set *NamedSet) Named
```

File: pkg/packetfilter/packetfilter.go [hunk 3]
```diff
@@ -153,6 +182,11 @@ type Driver interface {
...
+	DestroySet(set string) error
```
Comment (tpantelis) [inline pkg/packetfilter/packetfilter.go]: We really don't need this method. It's used in [one place](https://github.com/submariner-io/submariner/blob/636b457da21542ec3a7632143562c220c57e4f55/pkg/globalnet/controllers/uninstall.go#L106) to destroy all sets with a name prefix but the consumer can create a `NamedSet` and call `Destroy`. But I think a better alternative is to replace `ListSets` with a general `DestroySets` that takes a name filter:

`DestroySets(nameFilter func(string) bool) error`

Ideally `ListSets` would return a `NamedSet` but the ipset lib doesn't have a way to retrieve a set's params after creation so perhaps it's best not to expose just set names.

File: pkg/packetfilter/iptables/iptables.go [hunk 4]
```diff
@@ -52,7 +54,8 @@ var (
 )
 
 type PacketFilter struct {
-	ipt *iptables.IPTables
+	ipt        *iptables.IPTables
+	IPSetIface ipset.Interface
```
Comment (tpantelis) [inline pkg/packetfilter/iptables/iptables.go]: Let's not export this.

```suggestion
	ipSetIface ipset.Interface
```

File: pkg/packetfilter/iptables/namedset.go [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+	hashFamily := ipset.ProtocolFamilyIPV4
+	if set.Family == packetfilter.SetFamilyV6 {
+		hashFamily = ipset.ProtocolFamilyIPV6
+	}
+
+	return ipset.NewNamed(&ipset.IPSet{
+		Name:       set.Name,
+		SetType:    ipset.HashNet,
+		HashFamily: hashFamily,
+	}, a.IPSetIface)
```
Comment (tpantelis) [inline pkg/packetfilter/iptables/namedset.go]: I think we should move the `Named` implementation in ipset here as we expect consumers to use the API in packetfilter. We could move all the ipset code here as well but I think it makes sense to keep that in its own package as we inherited it from somewhere else.

File: pkg/packetfilter/fake/namedset.go [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+	"strings"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"github.com/submariner-io/submariner/pkg/packetfilter"
+	"k8s.io/utils/set"
+)
+
+type setInterface interface {
```
Comment (tpantelis) [inline pkg/packetfilter/fake/namedset.go]: I'm not clear on the purpose of this local interface. This seems to be patterned after `ipset.Interface` but we don't need all these methods.

File: pkg/packetfilter/fake/namedset.go [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+	TestEntry(entry string, set string) (bool, error)
+	// ListEntries lists all the entries from a named set
+	ListEntries(set string) ([]string, error)
+	// ListSets list all set names from kernel
+	ListSets() ([]string, error)
+}
+
+type namedType struct {
+	nSet  *packetfilter.NamedSet
+	iface setInterface
```
Comment (tpantelis) [inline pkg/packetfilter/fake/namedset.go]: Just use `PacketFilter` directly.

```suggestion
	pfilter *PacketFilter
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

File: pkg/packetfilter/fake/namedset.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+	"strings"
+	"time"
+
+	. "github.com/onsi/gomega"
+	"github.com/pkg/errors"
+	"github.com/submariner-io/submariner/pkg/packetfilter"
+	"k8s.io/utils/set"
+)
+
+type namedType struct {
```
Comment (tpantelis) [inline pkg/packetfilter/fake/namedset.go]: Minor thing but more clear.

```suggestion
type namedSet struct {
```

File: pkg/packetfilter/fake/namedset.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+	return n.setInfo.Name
+}
+
+func (n *namedType) Flush() error {
+	n.pfilter.flushSet(n.setInfo.Name)
+	return nil
+}
+
+func (n *namedType) Destroy() error {
+	return errors.Wrap(n.pfilter.destroySet(n.setInfo.Name), "Destroy failed")
```
Comment (tpantelis) [inline pkg/packetfilter/fake/namedset.go]: Minor but we should need to wrap here nor in the methods below.

```suggestion
	return n.pfilter.destroySet(n.setInfo.Name)
```

File: pkg/packetfilter/fake/namedset.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+		return err
+	}
+
+	if i.sets[setName] == nil {
+		return nil
+	}
+
+	delete(i.sets, setName)
+
+	return nil
```
Comment (tpantelis) [inline pkg/packetfilter/fake/namedset.go]: `delete` handles nil

```suggestion
	delete(i.sets, setName)

	return nil
```

File: pkg/packetfilter/fake/namedset.go [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+	for _, setName := range names {
+		if nameFilter(setName) {
+			err := matchForError(&i.failOnDestroySetMatchers, setName)
+			if err != nil {
+				return err
+			}
+
+			if i.sets[setName] != nil {
+				delete(i.sets, setName)
+			}
```
Comment (tpantelis) [inline pkg/packetfilter/fake/namedset.go]:
```suggestion
			delete(i.sets, setName)
```

File: pkg/packetfilter/iptables/namedset.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+
+package iptables
+
+import (
+	"github.com/pkg/errors"
+	"github.com/submariner-io/submariner/pkg/ipset"
+	"github.com/submariner-io/submariner/pkg/packetfilter"
+)
+
+type namedType struct {
```
Comment (tpantelis) [inline pkg/packetfilter/iptables/namedset.go]:
```suggestion
type namedSet struct {
```

**Review by tpantelis — CHANGES_REQUESTED** [review]

**Review by tpantelis — APPROVED** [review]

**Comment by yboaron** [comment]
cc @sridhargaddam

File: pkg/packetfilter/fake/namedset.go (hunk 8 above)
Comment (sridhargaddam) [reply pkg/packetfilter/fake/namedset.go]: Its just a minor thing, but just want to check if you wanted to use namedSet here? (as the conversation is marked as resolved).

File: pkg/packetfilter/fake/namedset.go [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+	defer i.mutex.Unlock()
+
+	err := matchForError(&i.failOnAddEntryMatchers, entry)
+	if err != nil {
+		return err
+	}
+
+	entries := i.sets[nSet.Name]
+	if entries == nil {
+		return fmt.Errorf("IP set %q does not exist", nSet.Name)
```
Comment (sridhargaddam) [inline pkg/packetfilter/fake/namedset.go]:
```suggestion
		return fmt.Errorf("named set %q does not exist", nSet.Name)
```

File: pkg/packetfilter/fake/namedset.go (hunk 8 above)
Comment (yboaron) [reply pkg/packetfilter/fake/namedset.go]: > Its just a minor thing, but just want to check if you wanted to use namedSet here? (as the conversation is marked as resolved).

Ooops,   missed it. 
thanks

**Review by sridhargaddam — APPROVED** [review]

**Pull request merged**